File: src/main/java/editor/gui/ccp/data/EntryTransferData.java
package editor.gui.ccp.data;

import java.awt.datatransfer.DataFlavor;
import java.awt.datatransfer.UnsupportedFlavorException;
import java.util.Arrays;
import java.util.Collections;
import java.util.Map;

import editor.database.card.Card;
import editor.gui.editor.EditorFrame;

/**
 * This class represents data being transferred via drag and drop or cut/copy/paste
 * between a CardList and another object. CardLists only supports importing card or entry
 * data flavors, but can export Strings as well.
 *
 * @author Alec Roelke
 */
public class EntryTransferData extends CardTransferData
{
    /** Cards being transferred. */
    public final Map<Card, Integer> entries;
    /** Frame containing the deck being transferred from. */
    public final EditorFrame source;
    /** Frame containing the deck being transferred to. */
    public EditorFrame target;
    /** ID of the list in the deck being transferred from. */
    public final int from;
    /** ID of the list int he deck being transferred to. */
    public int to;

    public EntryTransferData(EditorFrame e, int id, Map<Card, Integer> cards)
    {
        super(cards.keySet().stream().sorted(Card::compareName).toArray(Card[]::new));
        this.entries = Collections.unmodifiableMap(cards);
        source = e;
        target = null;
        from = id;
        to = -1;
    }

    @Override
    public Object getTransferData(DataFlavor flavor) throws UnsupportedFlavorException
    {
        if (flavor.equals(DataFlavors.entryFlavor))
            return this;
        else
            return super.getTransferData(flavor);
    }

    @Override
    public DataFlavor[] getTransferDataFlavors()
    {
        DataFlavor[] superFlavors = super.getTransferDataFlavors();
        DataFlavor[] flavors = new DataFlavor[superFlavors.length + 1];
        flavors[0] = DataFlavors.entryFlavor;
        for (int i = 0; i < superFlavors.length; i++)
            flavors[i + 1] = superFlavors[i];
        return flavors;
    }

    @Override
    public boolean isDataFlavorSupported(DataFlavor flavor)
    {
        return Arrays.asList(getTransferDataFlavors()).contains(flavor);
    }
}

File: src/main/scala/editor/gui/ccp/data/EntryTransferData.scala
package editor.gui.ccp.data

import editor.database.card.Card
import editor.gui.editor.EditorFrame

import java.awt.datatransfer.DataFlavor
import java.awt.datatransfer.UnsupportedFlavorException

/**
 * Cards being transferred from a card list to another element.
 * 
 * @constructor create a new set of transfer data for a specific list within a frame
 * @param source source frame containing the list to transfer from
 * @param from ID of the list to transfer from
 * @param cards cards and counts to transfer
 * 
 * @author Alec Roelke
 */
class EntryTransferData(val source: EditorFrame, val from: Int, val cards: Map[Card, Int]) extends CardTransferData(cards.keys.toSeq.sortBy(_.name).toArray) {
  /** Target frame containing the list to transfer to. */
  var target: EditorFrame = null
  /** ID of the list to transfer to. */
  var to = -1

  @throws[UnsupportedFlavorException]
  override def getTransferData(flavor: DataFlavor) = if (flavor == DataFlavors.entryFlavor) this else super.getTransferData(flavor)
  override def getTransferDataFlavors = DataFlavors.entryFlavor +: super.getTransferDataFlavors
  override def isDataFlavorSupported(flavor: DataFlavor) = getTransferDataFlavors.contains(flavor)
}